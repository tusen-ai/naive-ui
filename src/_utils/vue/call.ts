type AnyFunction = (...args: any[]) => any

function call (funcs: MaybeArray<() => void>): void
function call<A1> (funcs: MaybeArray<(a1: A1) => void>, a1: A1): void
function call<A1, A2> (
  funcs: MaybeArray<(a1: A1, a2: A2) => void>,
  a1: A1,
  a2: A2
): void
function call<A1, A2, A3> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3) => void>,
  a1: A1,
  a2: A2,
  a3: A3
): void
function call<A1, A2, A3, A4> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3, a4: A4) => void>,
  a1: A1,
  a2: A2,
  a3: A3,
  a4: A4
): void
function call<A extends any[]> (
  funcs: AnyFunction[] | AnyFunction,
  ...args: A
): void {
  if (Array.isArray(funcs)) {
    funcs.forEach((func) => (call as any)(func, ...args))
  } else return funcs(...args)
}

export { call }

export type MaybeArray<T> = T | T[]
