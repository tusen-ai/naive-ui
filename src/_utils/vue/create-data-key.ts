export function createDataKey(key: string | number): string {
  return typeof key === 'string' ? `s-${key}` : `n-${key}`
}
