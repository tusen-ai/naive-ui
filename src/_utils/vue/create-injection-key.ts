import type { InjectionKey } from 'vue'

export function createInjectionKey<T>(key: string): InjectionKey<T> {
  return key as any
}
