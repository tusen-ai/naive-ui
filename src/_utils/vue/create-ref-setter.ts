import type { Ref } from 'vue'

export function createRefSetter (ref: Ref<HTMLElement | null>): any {
  return (inst: { $el: HTMLElement | null } | null) => {
    if (inst) {
      ref.value = inst.$el
    } else {
      ref.value = null
    }
  }
}
