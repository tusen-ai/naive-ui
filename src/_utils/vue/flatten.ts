import {
  Fragment,
  type VNodeChild,
  createTextVNode,
  type VNode,
  Comment
} from 'vue'

// o(n) flatten
export function flatten (
  vNodes: VNodeChild[],
  filterCommentNode: boolean = true,
  result: VNode[] = []
): VNode[] {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      if (typeof vNode === 'string' || typeof vNode === 'number') {
        result.push(createTextVNode(String(vNode)))
      }
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, filterCommentNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, filterCommentNode, result)
      }
      // rawSlot
    } else if (vNode.type !== Comment) {
      result.push(vNode)
    }
  })
  return result
}
