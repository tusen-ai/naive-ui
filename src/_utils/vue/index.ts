export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
