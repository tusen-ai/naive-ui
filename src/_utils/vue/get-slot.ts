import { type ComponentPublicInstance, type VNodeChild } from 'vue'

export function getSlot (
  instance: ComponentPublicInstance,
  slotName = 'default',
  fallback: VNodeChild[] = []
): VNodeChild[] {
  const slots = instance.$slots
  const slot = slots[slotName]
  if (slot === undefined) return fallback
  return slot()
}
