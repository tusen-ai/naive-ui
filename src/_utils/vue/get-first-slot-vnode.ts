import { type Slots, type VNode } from 'vue'
import { flatten } from './flatten'
import { warn } from '../naive'

export function getFirstSlotVNode (
  slots: Slots,
  slotName = 'default',
  props: unknown = undefined
): VNode | null {
  const slot = slots[slotName]
  if (!slot) {
    warn('getFirstSlotVNode', `slot[${slotName}] is empty`)
    return null
  }
  const slotContent = flatten(slot(props))
  // vue will normalize the slot, so slot must be an array
  if (slotContent.length === 1) {
    return slotContent[0]
  } else {
    warn('getFirstSlotVNode', `slot[${slotName}] should have exactly one child`)
    return null
  }
}
