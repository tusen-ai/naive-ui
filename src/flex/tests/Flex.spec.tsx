import { mount, type VueWrapper } from '@vue/test-utils'
import { h, Fragment, createCommentVNode } from 'vue'
import { NFlex } from '../index'

const getChildrenNode = (wrapper: VueWrapper<any>): any[] => {
  return Array.from(wrapper.find('.n-flex').element.childNodes)
}

describe('n-flex', () => {
  it('should work with import on demand', () => {
    mount(NFlex)
  })

  it('render empty children', () => {
    const wrapper = mount({
      render () {
        return <NFlex />
      }
    })
    expect(wrapper.find('.n-flex')).not.toBe(null)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('render flex string size', () => {
    const wrapper = mount({
      render () {
        return <NFlex size="large">{{ default: () => 'kirby' }}</NFlex>
      }
    })
    expect(wrapper.attributes('style')).toContain('gap')
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('render vertical flex', () => {
    const wrapper = mount({
      render () {
        return (
          <NFlex vertical>
            {{
              default: () => [<div>1</div>, <div>2</div>]
            }}
          </NFlex>
        )
      }
    })
    expect(wrapper.attributes('style')).toContain('flex-direction: column;')
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('should work with `inline` prop', () => {
    const wrapper = mount(NFlex, {
      props: {
        inline: true
      },
      slots: {
        default: () => '07akioni'
      }
    })

    expect(wrapper.attributes('style')).toContain('display: inline-flex;')
    wrapper.unmount()
  })

  it('should render with invalidElement', () => {
    const wrapper = mount({
      render () {
        return (
          <NFlex>
            {{
              default: () => (
                <>
                  text1<span>text1</span>
                  text1
                </>
              )
            }}
          </NFlex>
        )
      }
    })

    const childNodes = getChildrenNode(wrapper)
    expect(childNodes.length).toBe(3)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('should not render comment node', () => {
    const wrapper = mount({
      render () {
        return <NFlex>{{ default: () => false && 'div' }}</NFlex>
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    const childNodes = getChildrenNode(wrapper)
    expect(childNodes.length).toEqual(1) // Comment node
  })

  it('should work with `wrap` prop', async () => {
    const wrapper = mount(NFlex, {
      slots: { default: () => [<div>1</div>, <div>2</div>] }
    })
    expect(wrapper.find('.n-flex').attributes('style')).toContain(
      'flex-wrap: wrap'
    )

    await wrapper.setProps({ wrap: false })
    expect(wrapper.find('.n-flex').attributes('style')).toContain(
      'flex-wrap: nowrap'
    )
    wrapper.unmount()
  })

  it('should render while slot is Comment', () => {
    const wrapper = mount({
      render () {
        return (
          <NFlex>
            {{
              default: () => createCommentVNode('random comment text')
            }}
          </NFlex>
        )
      }
    })
    const childNodes = getChildrenNode(wrapper)
    expect(childNodes.length).toEqual(1)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })
})
