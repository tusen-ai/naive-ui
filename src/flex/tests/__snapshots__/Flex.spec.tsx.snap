// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`n-flex render empty children 1`] = `""`;

exports[`n-flex render flex string size 1`] = `"<div role="none" class="n-flex" style="display: flex; flex-direction: row; justify-content: start; flex-wrap: wrap; gap: 12px 16px;">kirby</div>"`;

exports[`n-flex render vertical flex 1`] = `
"<div role="none" class="n-flex" style="display: flex; flex-direction: column; justify-content: start; flex-wrap: nowrap; gap: 8px 12px;">
  <div>1</div>
  <div>2</div>
</div>"
`;

exports[`n-flex should not render comment node 1`] = `
"<div role="none" class="n-flex" style="display: flex; flex-direction: row; justify-content: start; flex-wrap: wrap; gap: 8px 12px;">
  <!---->
</div>"
`;

exports[`n-flex should render while slot is Comment 1`] = `
"<div role="none" class="n-flex" style="display: flex; flex-direction: row; justify-content: start; flex-wrap: wrap; gap: 8px 12px;">
  <!--random comment text-->
</div>"
`;

exports[`n-flex should render with invalidElement 1`] = `"<div role="none" class="n-flex" style="display: flex; flex-direction: row; justify-content: start; flex-wrap: wrap; gap: 8px 12px;">text1<span>text1</span>text1</div>"`;
